src/translate.rs

```diff
@@ -1081,56 +1173,113 @@ impl FromGlib<glib_sys::gboolean> for bool {
     fn from_glib(val: i32) -> Ordering {
         val.cmp(&0)
     }
 }
 
-impl FromGlib<u32> for Option<char> {
+/// Translate from a Glib type which can result in an undefined and/or invalid value.
+pub trait TryFromGlib<G>: Sized {
+    type Error;
+    fn try_from_glib(val: G) -> Result<Self, Self::Error>;
+}
+
+/// Error type for `TryFromGlib` when the Glib value is None.
+#[derive(Debug, PartialEq, Eq)]
+pub struct NoneError;
+
+impl fmt::Display for NoneError {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(fmt, "value is None")
+    }
+}
+
+impl std::error::Error for NoneError {}
+
+impl<G, T: TryFromGlib<G, Error = NoneError>> FromGlib<G> for Option<T> {
     #[inline]
-    fn from_glib(val: u32) -> Option<char> {
-        match val {
-            0 => None,
-            _ => char::from_u32(val),
+    fn from_glib(val: G) -> Option<T> {
+        T::try_from_glib(val).ok()
+    }
+}
+
+/// Error type for `TryFromGlib` when the Glib value can be None or invalid.
+#[derive(Debug)]
+pub enum NoneOrInvalidError<I> {
+    Invalid(I),
+    None,
+}
+
+impl<I> NoneOrInvalidError<I> {
+    /// Builds the `None` variante.
+    pub fn none() -> Self {
+        NoneOrInvalidError::None
+    }
+
+    /// Returns `true` if `self` is the `None` variante.
+    // FIXME `matches!` was introduced in rustc 1.42.0, current MSRV is 1.40.0
+    // FIXME uncomment when CI can upgrade to 1.47.1
+    //#[allow(clippy::match_like_matches_macro)]
+    pub fn is_none(&self) -> bool {
+        match self {
+            NoneOrInvalidError::None => true,
+            _ => false,
+        }
+    }
+
+    /// Returns `true` if `self` is the `Invalid` variante.
+    // FIXME `matches!` was introduced in rustc 1.42.0, current MSRV is 1.40.0
+    // FIXME uncomment when CI can upgrade to 1.47.1
+    //#[allow(clippy::match_like_matches_macro)]
+    pub fn is_invalid(&self) -> bool {
+        match self {
+            NoneOrInvalidError::None => false,
+            _ => true,
         }
     }
 }
 
-impl FromGlib<i32> for Option<u32> {
-    #[inline]
-    fn from_glib(val: i32) -> Option<u32> {
-        if val >= 0 {
-            Some(val as u32)
-        } else {
-            None
+impl<I: Error> From<I> for NoneOrInvalidError<I> {
+    fn from(invalid: I) -> Self {
+        NoneOrInvalidError::Invalid(invalid)
+    }
+}
+
+impl<I: Error> fmt::Display for NoneOrInvalidError<I> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            NoneOrInvalidError::Invalid(err) => fmt::Display::fmt(err, fmt),
+            NoneOrInvalidError::None => write!(fmt, "value is None"),
         }
     }
 }
 
-impl FromGlib<i64> for Option<u64> {
-    #[inline]
-    fn from_glib(val: i64) -> Option<u64> {
-        if val >= 0 {
-            Some(val as u64)
-        } else {
-            None
+impl<I: Error> Error for NoneOrInvalidError<I> {}
+
+impl<I: PartialEq> PartialEq for NoneOrInvalidError<I> {
```

Comment:
Isn't this the same as the derived `PartialEq`?

Reply:
And `Eq` below

Reply:
The intention was to only implement these if `I` also implemented them. I didn't want to force those traits on `I`.

Reply:
That's what the default derive for them is doing :)

Reply:
Oh, I didn't know that! :) I thought it would have forced them on `I`. Thanks!